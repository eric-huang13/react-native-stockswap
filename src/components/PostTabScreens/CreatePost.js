import React, {Component} from 'react';
import {connect} from 'react-redux';
import {
  Text,
  View,
  StyleSheet,
  TextInput,
  SafeAreaView,
  KeyboardAvoidingView,
  Keyboard,
  TouchableWithoutFeedback,
  Switch,
  TouchableOpacity,
  ScrollView,
  Platform,
  Image,
} from 'react-native';
import {UserPost} from '../../actions/posts';
import {Formik} from 'formik';
import * as Yup from 'yup';
import Toast from 'react-native-toast-message';
import ImagePicker from 'react-native-image-crop-picker';

import {moderateScale} from '../../util/responsiveFont';

const validationSchema = Yup.object().shape({
  body: Yup.string()
    .label('body')
    .min(2, 'body must have more than 2 characters '),
});
class CreatePost extends Component {
  constructor(props) {
    super(props);

    this.state = {
      enabled: false,
    };
  }

  toggleSwitch = (value) => {
    this.setState({enabled: value});
  };
  //Creating FormData for sending image to backend
  createFormData = (values) => {
    let formData = new FormData();
    Object.keys(values).forEach((fieldName) => {
      console.log(fieldName, values[fieldName]);
      formData.append(fieldName, values[fieldName]);
    });
    return formData;
  };

  onImageSelection = (setFieldValue) => {
    ImagePicker.openPicker({
      forceJpg: true,
      mediaType: 'image',
      includeBase64: true,
      cropping: true,
    }).then((image) => {
      // console.log('Image:', image);
      const path = image?.path || image?.sourceURL;
      // console.log('Path:', path);
      if (path && setFieldValue) {
        setFieldValue('image', {
          name: image.filename,
          type: image.mime,
          data: image.data,
          uri: Platform.OS === 'android' ? path : path.replace('file://', ''),
        });
      }
    });
  };

  render() {
    //for testing
    // const credentials = {
    //       name: this.state.body,
    //       description:this.state.image,
    //       };
    const {UserPost} = this.props;

    return (
      <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
        <SafeAreaView style={style.mainContainer}>
          <ScrollView>
            <KeyboardAvoidingView
              behavior={Platform.OS === 'ios' ? 'padding' : null}
              style={{flex: 1}}>
              <Formik
                initialValues={{
                  enabled: false,
                  image: {name: '', type: '', uri: ''},
                  body: '',
                }}
                onSubmit={(values) => {
                  console.log(values, 'values');
                  const data = this.createFormData(values);
                  console.log(data, 'form');
                  // UserPost(values)
                }}
                validationSchema={validationSchema}>
                {({
                  handleChange,
                  values,
                  handleSubmit,
                  errors,
                  isValid,
                  touched,
                  handleBlur,
                  isSubmitting,
                  setFieldValue,
                }) => (
                  <View>
                    <Text style={style.header}> Post a Publication </Text>
                    {values.image.uri && !errors.image ? (
                      <TouchableOpacity
                        onPress={() => this.onImageSelection(setFieldValue)}>
                        <Image
                          style={style.uploadImageContainer}
                          source={{uri: values.image.uri}}
                        />
                      </TouchableOpacity>
                    ) : (
                      <View style={style.uploadImageContainer}>
                        <TouchableOpacity
                          onPress={() => this.onImageSelection(setFieldValue)}>
                          <Text style={style.uploadImageText}>
                            Tap to upload your photo
                          </Text>
                        </TouchableOpacity>
                      </View>
                    )}
                    {/* <View style={style.uploadImageContainer}>
              <TextInput
                onBlur={handleBlur('image')}
                value={values.image}
                onChangeText={handleChange('image')}
                placeholder="Upload cover image"
                placeholderTextColor="#FFFFFF"
                style={style.inputStyleImage}
              />
              <Text style={style.errorText}>
                          {touched.image && errors.image}
                        </Text>
            </View> */}
                    <View style={style.postContainer}>
                      <Text style={style.inputHeader}>Post</Text>
                      <TextInput
                        style={style.inputStyleBody}
                        // onBlur={handleBlur('body')}
                        value={values.body}
                        onChangeText={handleChange('body')}
                        placeholder="Enter post text"
                        placeholderTextColor="#9ea6b5"
                        multiline={true}
                        numberOfLines={4}
                      />
                      <Text style={style.errorText}>{errors.body}</Text>
                    </View>

                    <View style={style.notificationsContainer}>
                      <Text style={style.middleDetailsText}>
                        Turn off comments
                      </Text>
                      <Switch
                        trackColor={{false: '#1A2542', true: '#B8A0FF'}}
                        thumbColor={this.state.enabled ? '#f4f3f4' : '#f4f3f4'}
                        ios_backgroundColor="#f4f3f4"
                        onValueChange={(value) =>
                          setFieldValue('enabled', value)
                        }
                        value={values.enabled}
                        style={{
                          transform: [
                            {scaleX: moderateScale(1.5)},
                            {scaleY: moderateScale(1.5)},
                          ],
                        }}
                      />
                    </View>
                    <View style={style.buttonsContainer}>
                      <TouchableOpacity onPress={() => handleSubmit()}>
                        <Text style={style.publishButton}>Publish</Text>
                      </TouchableOpacity>
                      <TouchableOpacity
                        onPress={() =>
                          values.body === '' && values.image === ''
                            ? Toast.show({
                                type: 'error',
                                text2: 'Post must include body or image.',
                              })
                            : errors.body || errors.image
                            ? Toast.show({
                                type: 'error',
                                text2: 'Please fix errors',
                              })
                            : this.props.navigation.navigate({
                                name: 'CreatePostPreview',
                                params: {data: values},
                              })
                        }>
                        <Text style={style.previewButton}>Preview</Text>
                      </TouchableOpacity>
                    </View>
                  </View>
                )}
              </Formik>
            </KeyboardAvoidingView>
          </ScrollView>
        </SafeAreaView>
      </TouchableWithoutFeedback>
    );
  }
}
const mapStateToProps = (state) => {
  return {
    userAccount: state.user.userFakeData,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    UserPost: (input) => dispatch(UserPost(input)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(CreatePost);
const style = StyleSheet.create({
  mainContainer: {
    flex: 1,
    backgroundColor: '#2a334a',
    paddingVertical: moderateScale(20),
    paddingHorizontal: moderateScale(26),
  },
  header: {
    marginTop: moderateScale(20),
    fontSize: moderateScale(20),
    fontFamily: 'Montserrat-Bold',
    color: '#FFFFFF',
    textAlign: 'center',
  },
  uploadImageContainer: {
    marginTop: moderateScale(20),
    backgroundColor: '#46486e',
    width: '100%',
    alignSelf: 'center',
    height: moderateScale(130),
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: moderateScale(2),
  },
  uploadImageText: {
    fontSize: moderateScale(14),
    color: '#FFFFFF',
    fontFamily: 'Montserrat-Regular',
  },
  inputStyleImage: {
    fontFamily: 'Montserrat-Regular',
    color: '#FFFFFF',
    width: moderateScale(200),
    textAlign: 'center',
  },
  postContainer: {
    width: '100%',
    alignSelf: 'center',
    marginTop: moderateScale(20),
  },
  inputHeader: {
    fontSize: moderateScale(12),
    color: '#babec8',
    marginBottom: moderateScale(1),
    fontFamily: 'Montserrat-Regular',
  },
  inputStyleBody: {
    borderRadius: moderateScale(6),
    backgroundColor: '#3e4d6c',
    marginBottom: moderateScale(12),
    padding: moderateScale(8),
    marginTop: moderateScale(1),
    fontSize: moderateScale(14),
    textAlignVertical: 'top',
    fontFamily: 'Montserrat-Italic',
    color: '#FFFFFF',
    height: moderateScale(200),
  },
  notificationsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: moderateScale(10),
    marginBottom: moderateScale(10),
    paddingRight: moderateScale(16),
  },
  middleDetailsText: {
    fontFamily: 'Montserrat-SemiBold',
    color: '#FFFFFF',
    fontSize: moderateScale(16),
  },
  buttonsContainer: {
    marginTop: moderateScale(40),
    flexDirection: 'row',
    justifyContent: 'space-between',
    // marginHorizontal:30,
  },
  publishButton: {
    alignSelf: 'center',
    color: '#8b64ff',
    textAlign: 'center',
    paddingVertical: moderateScale(12),
    paddingHorizontal: moderateScale(20),
    width: moderateScale(160),
    borderRadius: moderateScale(6),
    fontSize: moderateScale(14),
    fontFamily: 'Montserrat-SemiBold',
    borderWidth: moderateScale(1),
    borderColor: '#8b64ff',
  },
  previewButton: {
    alignSelf: 'center',
    backgroundColor: '#8b64ff',
    color: '#FFFFFF',
    textAlign: 'center',
    paddingVertical: moderateScale(12),
    paddingHorizontal: moderateScale(20),
    width: moderateScale(160),
    borderRadius: moderateScale(6),
    fontSize: moderateScale(14),
    fontFamily: 'Montserrat-SemiBold',
  },
  errorText: {
    color: '#F66E6E',
    fontWeight: 'bold',
    marginBottom: moderateScale(1),
    marginTop: moderateScale(1),
    textAlign: 'center',
  },
});
