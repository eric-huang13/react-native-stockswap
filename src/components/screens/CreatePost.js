import React, { Component } from 'react'
import { Text, View, StyleSheet, TextInput, SafeAreaView, KeyboardAvoidingView, Keyboard, TouchableWithoutFeedback,Switch, TouchableOpacity } from 'react-native'


export default class CreatePost extends Component {
    constructor(props) {
        super(props);
    
        this.state = {
          enabled: false,
          image:'',
          body:'',
        };
      }
      
      toggleSwitch = (value) => {
        this.setState({enabled: value});
      };
      handleImageChange = (text) => {
        this.setState({
          image: text,
        });
      };
      handleBodyChange = (text) => {
        this.setState({
          body: text,
        });
      };
      
    render() {
        
        return (
        
        
              <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
            <SafeAreaView style={style.mainContainer}>
            <KeyboardAvoidingView
          behavior={Platform.OS === 'ios' ? 'padding' : null}
          style={{flex: 1}}>
                <Text style={style.header}> Post a Publication </Text>
                <View style={style.uploadImageContainer}>               
                  <TextInput
                    value={this.state.image}
                    onChangeText={(text) => this.handleImageChange(text)}
                    placeholder="Upload cover image"
                    placeholderTextColor="#FFFFFF"
                    style={style.inputStyleImage}
                    ref={(input) => (this.image = input)}
                    onSubmitEditing={() => this.hashtag.focus()}
                  />
                </View>
                <View style={style.postContainer}>
                  <Text style={style.inputHeader}>Post</Text>
                  <TextInput
                    style={style.inputStyleBody}
                    value={this.state.body}
                    onChangeText={(text) => this.handleBodyChange(text)}
                    placeholder="Enter post text"
                    placeholderTextColor="#9ea6b5"
                    multiline={true}
                    numberOfLines={4}
                    ref={(input) => (this.body = input)}
                  />
                </View>
                
                <View style={style.notificationsContainer}>
              <Text style={style.middleDetailsText}>
                Turn off comments
              </Text>
              <Switch
                trackColor={{false: '#1A2542', true: '#B8A0FF'}}
                thumbColor={this.state.enabled ? '#f4f3f4' : '#f4f3f4'}
                ios_backgroundColor="#f4f3f4"
                onValueChange={this.toggleSwitch}
                value={this.state.enabled}
                style={{transform: [{scaleX: 1.5}, {scaleY: 1.5}]}}
              />
            </View>
            <View style={style.buttonsContainer}>
                  <TouchableOpacity
                    onPress={() => this.props.navigation.goBack()}>
                    <Text style={style.publishButton}>Publish</Text>
                  </TouchableOpacity>
                  <TouchableOpacity
                onPress={() =>
                  this.props.navigation.navigate('CreatePostPreview')
                }>
                  <Text style={style.previewButton}>Preview</Text>
                </TouchableOpacity>
                  {/* <TouchableOpacity
                    onPress={() => EditUserAccount(this.state)}>
                    <Text style={style.previewButton}>Preview</Text>
                  </TouchableOpacity> */}
                </View>
                </KeyboardAvoidingView>
            </SafeAreaView>
            </TouchableWithoutFeedback>
       
            
        )
    }
}

const style = StyleSheet.create({
    mainContainer: {
        flex:1,
        backgroundColor:'#2a334a',
        paddingVertical:20, 
        paddingHorizontal:26,  
        },
    header:{
        marginTop:20,
        fontSize:20,
        fontFamily: 'Montserrat-Bold',
        color:'#FFFFFF', 
        textAlign:'center',   
    },
    uploadImageContainer:{
        marginTop:20,
        backgroundColor:'#46486e',
        width:'100%',
        alignSelf:'center',
        height:130,
        alignItems:'center',
        justifyContent:'center',
        borderRadius:2,
    },
    uploadImageText:{
        fontSize:14,
        color:'#FFFFFF',
        fontFamily: 'Montserrat-Regular',

    },
    inputStyleImage: {
        fontFamily: 'Montserrat-Regular',
        color: '#FFFFFF',
        width:200,
        textAlign:'center'
                
      },
    postContainer:{
        width:'100%',
        alignSelf:'center',
        marginTop:20,
    },
    inputHeader: {
        fontSize: 12,
        color: '#babec8',
        marginBottom: 1,
        fontFamily: 'Montserrat-Regular',
      },
      inputStyleBody: {
        borderRadius: 6,
        backgroundColor: '#3e4d6c',
        marginBottom: 12,
        padding: 8,
        marginTop: 1,
        fontSize: 14,
        textAlignVertical: 'top',
        fontFamily: 'Montserrat-Italic',
        color: '#FFFFFF',
        height:200,
        
      },
      notificationsContainer: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginTop: 10,
        marginBottom: 10,
        // marginHorizontal:30,
      },
      middleDetailsText: {
        fontFamily: 'Montserrat-SemiBold',
        color: '#FFFFFF',
        fontSize: 16,
      },
      buttonsContainer: {
        marginTop: 40,
        flexDirection: 'row',
        justifyContent: 'space-between',
        // marginHorizontal:30,
      },
      publishButton: {
        alignSelf: 'center',
        color: '#8b64ff',
        textAlign: 'center',
        paddingVertical: 12,
        paddingHorizontal: 20,
        width: 160,
        borderRadius: 6,
        fontSize: 14,
        fontFamily: 'Montserrat-SemiBold',
        borderWidth: 1,
        borderColor: '#8b64ff',
      },
      previewButton: {
        alignSelf: 'center',
        backgroundColor: '#8b64ff',
        color: '#FFFFFF',
        textAlign: 'center',
        paddingVertical: 12,
        paddingHorizontal: 20,
        width: 160,
        borderRadius: 6,
        fontSize: 14,
        fontFamily: 'Montserrat-SemiBold',
      },
})