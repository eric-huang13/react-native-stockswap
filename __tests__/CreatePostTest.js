import React from 'react';
import CreatePost from '../src/components/PostTabScreens/CreatePost';
import renderer from 'react-test-renderer';


jest.mock("@react-native-async-storage/async-storage", () =>
  require("@react-native-async-storage/async-storage/jest/async-storage-mock"),
);
jest.mock('react-redux', () => ({
    connect: () => {
      return (component) => {
        return component
      };
    }
  }));
  

test('renders correctly', () => {
    const tree = renderer.create(<CreatePost/>).toJSON();
  expect(tree).toMatchSnapshot();
});