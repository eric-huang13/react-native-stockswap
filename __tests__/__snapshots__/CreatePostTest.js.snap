// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  accessible={true}
  emulateUnlessSupported={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#2a334a",
      "flex": 1,
      "paddingHorizontal": 36.722627737226276,
      "paddingVertical": 28.248175182481752,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Montserrat-Bold",
                "fontSize": 28.248175182481752,
                "marginTop": 28.248175182481752,
                "textAlign": "center",
              }
            }
          >
             Post a Publication 
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#46486e",
                "borderRadius": 2.8248175182481754,
                "height": 183.61313868613138,
                "justifyContent": "center",
                "marginTop": 28.248175182481752,
                "width": "100%",
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Montserrat-Regular",
                    "fontSize": 19.77372262773723,
                  }
                }
              >
                Tap to upload your photo
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignSelf": "center",
                "marginTop": 28.248175182481752,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#babec8",
                  "fontFamily": "Montserrat-Regular",
                  "fontSize": 16.94890510948905,
                  "marginBottom": 1.4124087591240877,
                }
              }
            >
              Post
            </Text>
            <TextInput
              allowFontScaling={true}
              multiline={true}
              numberOfLines={4}
              onChangeText={[Function]}
              placeholder="Enter post text"
              placeholderTextColor="#9ea6b5"
              rejectResponderTermination={true}
              style={
                Object {
                  "backgroundColor": "#3e4d6c",
                  "borderRadius": 8.474452554744525,
                  "color": "#FFFFFF",
                  "fontFamily": "Montserrat-Italic",
                  "fontSize": 19.77372262773723,
                  "height": 282.4817518248175,
                  "marginBottom": 16.94890510948905,
                  "marginTop": 1.4124087591240877,
                  "padding": 11.299270072992702,
                  "textAlignVertical": "top",
                }
              }
              underlineColorAndroid="transparent"
              value=""
            />
            <Text
              style={
                Object {
                  "color": "#F66E6E",
                  "fontWeight": "bold",
                  "marginBottom": 1.4124087591240877,
                  "marginTop": 1.4124087591240877,
                  "textAlign": "center",
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 14.124087591240876,
                "marginTop": 14.124087591240876,
                "paddingRight": 22.598540145985403,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "Montserrat-SemiBold",
                  "fontSize": 22.598540145985403,
                }
              }
            >
              Turn off comments
            </Text>
            <RCTSwitch
              accessibilityRole="switch"
              onChange={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              onTintColor="#B8A0FF"
              style={
                Array [
                  Object {
                    "height": 31,
                    "width": 51,
                  },
                  Array [
                    Object {
                      "transform": Array [
                        Object {
                          "scaleX": 2.1186131386861313,
                        },
                        Object {
                          "scaleY": 2.1186131386861313,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "#f4f3f4",
                      "borderRadius": 16,
                    },
                  ],
                ]
              }
              thumbTintColor="#f4f3f4"
              tintColor="#1A2542"
              value={false}
            />
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginTop": 56.496350364963504,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignSelf": "center",
                    "borderColor": "#8b64ff",
                    "borderRadius": 8.474452554744525,
                    "borderWidth": 1.4124087591240877,
                    "color": "#8b64ff",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 19.77372262773723,
                    "paddingHorizontal": 28.248175182481752,
                    "paddingVertical": 16.94890510948905,
                    "textAlign": "center",
                    "width": 225.98540145985402,
                  }
                }
              >
                Publish
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "#8b64ff",
                    "borderRadius": 8.474452554744525,
                    "color": "#FFFFFF",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 19.77372262773723,
                    "paddingHorizontal": 28.248175182481752,
                    "paddingVertical": 16.94890510948905,
                    "textAlign": "center",
                    "width": 225.98540145985402,
                  }
                }
              >
                Preview
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
